ruma-appservice-api/src/lib.rs

```diff
@@ -4,6 +4,138 @@
 
 #![warn(missing_debug_implementations, missing_docs)]
 
+use serde::{Deserialize, Serialize};
+
 pub mod event;
 pub mod query;
 pub mod thirdparty;
+
+/// A namespace defined by an application service.
+///
+/// Used for [appservice registration](https://matrix.org/docs/spec/application_service/r0.1.2#registration).
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct Namespace {
+    /// Whether this application service has exclusive access to events within this namespace.
+    pub exclusive: bool,
+
+    /// A regular expression defining which values this namespace includes.
+    pub regex: String,
+}
+
+impl Namespace {
+    /// Creates a new `Namespace` with the given exclusivity and regex pattern.
+    pub fn new<S: AsRef<str>>(exclusive: bool, regex: S) -> Self {
```

Comment:
Sometimes `Into<String>` can be useful for this kind of thing but I'm not sure we should use that here. Let's just use `String`.